src/components/Plans/PlanCalculator/reducers/plansByTypeReducer.test.js

```diff
@@ -72,53 +72,232 @@ describe('plansByTypeReducer', () => {
+          selectedDiscountIndex: expectedSelectedDiscountIndex,
+        }),
+      );
+    });
+
+    it(`should set a new discount and reset discount position when index position does not exist`, () => {
+      // Arrange
+      // TODO: it is not a realistic test because selectedDiscount should be one of the
+      // plan's discounts
+
+      const discountA = {
+        id: 1,
+        subscriptionType: SUBSCRIPTION_TYPE.quarterly,
+        discountPercentage: 10,
+      };
+
+      const discountB = {
+        id: 2,
+        subscriptionType: SUBSCRIPTION_TYPE.monthly,
+        discountPercentage: 20,
+      };
+      const discountC = {
+        id: 3,
+        subscriptionType: SUBSCRIPTION_TYPE.yearly,
+        discountPercentage: 30,
+      };
+
+      const discounts = [discountA, discountB];
+
+      const expectedSelectedDiscountIndex = 0;
+
+      const initialState = {
+        ...INITIAL_STATE_PLANS_BY_TYPE,
+        discounts,
+      };
+
+      const action = {
+        type: PLANS_BY_TYPE_ACTIONS.SELECT_DISCOUNT,
+        payload: discountC,
+      };
+
+      // Act
+      const newState = plansByTypeReducer(initialState, action);
+
+      // Assert
+      expect(newState).toEqual(
+        expect.objectContaining({
+          ...initialState,
+          selectedDiscount: discountA,
+          selectedDiscountIndex: expectedSelectedDiscountIndex,
+        }),
+      );
     });
   });
 
-  it(`${PLANS_BY_TYPE_ACTIONS.SELECT_PLAN} action`, () => {
-    // Arrange
-    const desiredPlan = { desiredPlan: true };
-    const initialState = {
-      ...INITIAL_STATE_PLANS_BY_TYPE,
-      plansByType: [{}, {}, {}, {}, {}, desiredPlan, {}],
-    };
-    const selectedPlanIndex = 5;
-    const action = {
-      type: PLANS_BY_TYPE_ACTIONS.SELECT_PLAN,
-      payload: selectedPlanIndex,
-    };
+  describe(`${PLANS_BY_TYPE_ACTIONS.SELECT_PLAN} action`, () => {
+    it('Should select a new plan and set previous selected discount when its index exists in new plan discounts ', () => {
+      // Arrange
 
-    // Act
-    const newState = plansByTypeReducer(initialState, action);
+      const discountA = {
+        id: 1,
+        billingCycle: 'monthly',
+        discountPercentage: 10,
+      };
 
-    // Assert
-    expect(newState).toEqual(
-      expect.objectContaining({
-        selectedPlanIndex,
-        selectedPlan: desiredPlan,
-      }),
-    );
+      const discountB = {
+        id: 2,
+        billingCycle: 'quarterly',
+        discountPercentage: 20,
+      };
+      const discountC = {
+        id: 3,
+        billingCycle: 'yearly',
+        discountPercentage: 30,
+      };
+
+      const discounts = [discountA, discountB, discountC];
+      const desiredPlan = { desiredPlan: true, billingCycleDetails: discounts };
+      const initialState = {
+        ...INITIAL_STATE_PLANS_BY_TYPE,
+        plansByType: [{}, {}, {}, {}, {}, desiredPlan, {}],
+        selectedDiscountIndex: 2,
+        discounts,
```

Comment:
I think that this state is inconsistent, I think that `selectedDiscount` is missed, and probably something else.

```suggestion
      const initialState = {
        ...INITIAL_STATE_PLANS_BY_TYPE,
        plansByType: [{}, {}, {}, {}, {}, desiredPlan, {}],
        selectedDiscountIndex: 2,
        selectedDiscount: discountC,
        discounts,
```

Reply:
in this case selectedDiscount state was not necessary to be setted in initialState, because when you change the plan, that selectedDiscount belongs to previous selected plan. So, the state wich cares, it's selectedDiscountIndex. It lets me know in wich array index I have to stay positioned to get the discount of the new plan.